rfcs/OptInFeatures.md

```diff
@@ -0,0 +1,152 @@
+
+  
+### Goals
+The goal of this proposal is to support the `experimental` state and, moving forward, any state that requires the client developer to make an explicit decision before using a given feature. In that sense, it's about "opting in" to using the feature, which includes supporting `experimental` states. 
+
+To give a few examples:
+
+* a field is experimental and might be changed or removed without prior notice (the above example).
+
+* a field is expensive to compute and should be used with caution.
+
+* a field has specific security requirements and requires a specific header or other form of authentication.
+
+### Non-goals
+This proposal is not about security and/or hiding parts of a schema. Its goal is to make it easier to communicate opt-in features to client developer and therefore needs to expose that information.
+
+## 👀 Prior work
+
+* GitHub uses [schema previews](https://docs.github.com/en/graphql/overview/schema-previews) to opt-in new features.
+* Kotlin has [OptIn requirements](https://kotlinlang.org/docs/opt-in-requirements.html) that started out as `@Experimental` before [being changed to `@OptIn`](https://youtrack.jetbrains.com/issue/KT-26216/Generalize-Experimental-API)
+
+## 🧑‍💻 Proposed solution
+
+### The `@optIn` directive
+
+It is proposed to add an `@optIn` directive to the specification:
+
+```graphql
+"""
+Indicates that the given field, argument, input field or enum value requires
+giving explicit consent before being used.
+"""
+directive @optIn(feature: String!) repeatable
+    on FIELD_DEFINITION 
+    | ARGUMENT_DEFINITION 
+    | INPUT_FIELD_DEFINITION 
+    | ENUM_VALUE
+```
+
+The `optIn` directive can then be used in the schema. For an example, to signal an experimental field:
+
+```graphql
+type Session {
+  id: ID!
+  title: String!
+  # [...]
+  startInstant: Instant @optIn(feature: "experimentalInstantApi")
+  endInstant: Instant @optIn(feature: "experimentalInstantApi")
+}
+```
+
+### Introspection
+
+> This section is a proposal based on the current introspection mechanism. A more global mechanism (see [#300](https://github.com/graphql/graphql-spec/issues/300)) would make it obsolete
+
+`@optIn` features should be hidden from introspection by default and include if `includeOptIn` contains the given feature:
+
+```graphql
+type __Type {
+  kind: __TypeKind!
+  name: String
+
+  # [...] other fields omitted for clarity
+
+  # includeOptIn is a list of features to include
+  fields(includeDeprecated: Boolean = false, includeOptIn: [String!]): [__Field!]  
+}
+```
+
+Tools can get a list of `@optIn` features required to use a field (or input field, argument, enum value) using `requiresOptIn`:
```

Comment:
Do tools also need a mechanism for listing all of the `@optIn` features in a schema? I'm imagining tools like GraphiQL showing users a list of possible features they can opt into rather than requiring users to somehow obtain this information via documentation or some other non-introspection based mechanism.

Reply:
Good point 👍 If types are not included in introspection in the first place, there's no way of knowing what possible values are for `includeOptIn`...

That'd be an argument for [marker directives](https://github.com/graphql/graphql-wg/blob/6a47b782a5d88f83dd6004d9d4e9385e2e2ee940/rfcs/OptInFeatures.md#marker-directives) because these user-defined directive would be returned no matter what.

The other way to look at this would be to return all the fields all the time (no more `includeOptIn`) and let the tooling (GraphiQL, Playground, etc...) use the logic they find appropriate to show them in autocomplete (or hide them)? I feel like that'd be my favorite solution. I never really understood why some tools would set `includeDeprecated: false` for an example. Is there a place where there is more context about this?